rasa/utils/tensorflow/callback.py

```diff
@@ -86,15 +86,20 @@ def _does_model_improve(self, current_results: Dict[Text, Any]) -> bool:
                 self.best_metrics_so_far[key] = float(current_results[key])
             return True
 
-        all_improved = all(
-            [
-                float(current_results[key]) > self.best_metrics_so_far[key]
-                for key in self.best_metrics_so_far.keys()
-            ]
-        )
-
-        if all_improved:
+        at_least_one_improved = False
+        for key in self.best_metrics_so_far.keys():
+            if key not in current_results:
+                # current_results.keys is not a subset of best_metric_so_far.keys
+                return False
```

Comment:
Can this happen?

Reply:
Yeah, it happens all the time actually, but due to a weird design decision (not made by me), so I wonder if I should refactor it to be cleaner/clearer. Here's what's happening:
This method is called at the end of **every** epoch, but metrics are only prefixed with `val` if it's an epoch where we've specified validation (so if its `epoch_number%evaluate_every_number_of_epoch==0`). So basically we fail this condition and nothing happens.

Reply:
I'll refactor it, you asking this question is a good push :)

Reply:
Oh, that was a rather creative design decision, I must say 😆  Thanks a lot for taking care of it!